### PR TITLE
Add multitenant APIs and tenant index management logic

### PR DESCRIPTION
**Pull request by zengyan-amazon**
*Description of changes:* 
* Add multitenant APIs and tenant index management logic
* Address comments from https://github.com/opendistro-for-elasticsearch/security-kibana-plugin/pull/162#pullrequestreview-392622995

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: server/auth/types/basic/basic_auth.ts [hunk 1]
```diff
@@ -17,9 +17,10 @@ import { AuthenticationHandler, SessionStorageFactory, IRouter, IClusterClient }
 import { SecurityPluginConfigType } from "../../..";
 import { SecuritySessionCookie } from "../../../session/security_cookie";
 import { CoreSetup } from "../../../../../../src/core/server";
-import _ from 'lodash';
-import { SecurityClient } from "../../../backend/opendistro_security_client";
-import { BasicAuthRoutes } from "./routes";
+import { assign } from 'lodash';
```
Comment (seraphjiang) [inline server/auth/types/basic/basic_auth.ts:20]: we could use ES buildin ```Object.assign```, since node support it

File: server/auth/types/basic/basic_auth.ts [hunk 2]
```diff
@@ -94,22 +104,37 @@ export class BasicAuthentication {
       if (!cookie) {
```
Comment (seraphjiang) [inline server/auth/types/basic/basic_auth.ts:104]: looks like cookie never defined

File: server/auth/types/basic/basic_auth.ts [hunk 3]
```diff
@@ -94,22 +104,37 @@ export class BasicAuthentication {
...
+      let headers = {};
+
       // set cookie to extend ttl
       cookie.expiryTime = Date.now() + this.config.cookie.ttl;
       this.sessionStorageFactory.asScoped(request).set(cookie);
-      
+
       // pass credentials to request to Elasticsearch
       const credentials = cookie.credentials;
+      assign(headers, { authorization: credentials.authHeaderValue });
```
Comment (seraphjiang) [inline server/auth/types/basic/basic_auth.ts:116]: do we need to test null for credentials?

File: server/auth/types/basic/routes.ts [hunk 4]
```diff
@@ -122,10 +77,10 @@ export class BasicAuthRoutes {
           authType: 'basicauth',
           isAnonymousAuth: false,
           expiryTime: Date.now() + this.config.cookie.ttl,
-        } 
+        };
         this.sessionStorageFactory.asScoped(request).set(sessionStorage);
 
-        if (this.config.multitenancy.enabled) {
+        if (this.config.multitenancy?.enabled || false) {
```
Comment (seraphjiang) [inline server/auth/types/basic/routes.ts:83]: why test ``` || false```, seem not necessary

File: server/multitenancy/tenant_resolver.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+ *   permissions and limitations under the License.
+ */
+
+import { KibanaRequest } from '../../../../src/core/server';
+import { SecurityPluginConfigType } from '..';
+import { SecuritySessionCookie } from '../session/security_cookie';
+import { isEmpty, findKey, cloneDeep } from 'lodash';
+
+/**
+ * Resovles the tenant the user is using.
```
Comment (seraphjiang) [inline server/multitenancy/tenant_resolver.ts:22]: is this jsdoc for resolveTenat() method below, if so should we move to above the method?

File: server/multitenancy/tenant_resolver.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+export function resolveTenant(
+  request: KibanaRequest,
+  username: string,
+  availabeTenants: any,
+  config: SecurityPluginConfigType,
+  cookie: SecuritySessionCookie
+): string {
+  let selectedTenant: string = undefined;
+  if (request.query && ((request.query as any).security_tenant || (request.query as any).securitytenant)) {
```
Comment (seraphjiang) [inline server/multitenancy/tenant_resolver.ts:44]: request.query as any has been used many time
could we define ```const query: any = request.query to improve readability

File: server/multitenancy/tenant_resolver.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      ? (request.query as any).security_tenant
+      : (request.query as any).securitytenant;
+  } else if (request.headers.securitytenant || request.headers.security_tenant) {
+    selectedTenant = request.headers.securitytenant
+      ? (request.headers.securitytenant as string)
+      : (request.headers.security_tenant as string);
+  } else if (cookie.tentent) {
+    selectedTenant = cookie.tentent;
+  } else {
+    selectedTenant = undefined;
```
Comment (seraphjiang) [inline server/multitenancy/tenant_resolver.ts:55]: no need this, already covered by initialization.

File: server/multitenancy/tenant_resolver.ts [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+function resolve(
+  username: string,
+  requestedTenant: string,
+  preferredTenants: string[],
+  availableTenants: any, // is an object like { tenant_name_1: true, tenant_name_2: false, ... }
+  globalTenantEnabled: boolean,
+  privateTenantEnabled: boolean
+): string {
+  let availableTenantsClone = cloneDeep(availableTenants);
```
Comment (seraphjiang) [inline server/multitenancy/tenant_resolver.ts:88]: should be fine to use const here, we could still delete property from const obj

File: server/routes/test_routes.ts [hunk 9]
```diff
@@ -62,6 +62,7 @@ export function defineTestRoutes(router: IRouter,
         }
       });
     } catch (error) {
+      console.log(error);
```
Comment (seraphjiang) [inline server/routes/test_routes.ts:65]: does error here contains username, password?

btw, do we still need a test_routes.ts?

File: server/routes/index.ts [hunk 10]
```diff
@@ -16,7 +16,30 @@
 import { IRouter, IClusterClient } from '../../../../src/core/server';
 import { schema } from '@kbn/config-schema';
 
-export function defineRoutes(router: IRouter, securityConfigClient: IClusterClient) {
+// TODO: conside to extract entity CRUD operations and put it into a client class
```
Comment (yuxitang-amzn) [inline server/routes/index.ts:19]: typo

File: server/routes/index.ts [hunk 11]
```diff
@@ -59,30 +82,30 @@ export function defineRoutes(router: IRouter, securityConfigClient: IClusterClie
...
+        inputSchema = roleSchema;
+        break;
+      case 'tenants':
+        inputSchema = tenantSchema;
+        break;
+      case 'account':
+        inputSchema = accountSchema;
+        break;
+      default:
+        throw new Error(`Unknown resource ${resourceName}`);
```
Comment (yuxitang-amzn) [inline server/routes/index.ts:104]: I think use a dict like below is more readable.
~~~
schemaMap = {
  "internalusers": internalUserSchema, 
  "actiongroups": actionGroupSchema,
  ...
}
inputSchema = schemaMap[resourceName]
if (inputSchema == undefined)
  throw
~~~

File: server/auth/types/basic/basic_auth.ts (hunk 3 above)
Comment (zengyan-amazon) [reply server/auth/types/basic/basic_auth.ts]: changed to optional chaining:
```
Object.assign(headers, { authorization: cookie.credentials?.authHeaderValue });
```

File: server/auth/types/basic/routes.ts (hunk 4 above)
Comment (zengyan-amazon) [reply server/auth/types/basic/routes.ts]: yes removed `|| false`

File: server/multitenancy/tenant_resolver.ts (hunk 5 above)
Comment (zengyan-amazon) [reply server/multitenancy/tenant_resolver.ts]: ah, yes, good catch

File: server/multitenancy/tenant_resolver.ts (hunk 6 above)
Comment (zengyan-amazon) [reply server/multitenancy/tenant_resolver.ts]: updated accordingly

File: server/multitenancy/tenant_resolver.ts (hunk 8 above)
Comment (zengyan-amazon) [reply server/multitenancy/tenant_resolver.ts]: updated to const

File: server/routes/test_routes.ts (hunk 9 above)
Comment (zengyan-amazon) [reply server/routes/test_routes.ts]: this API is for my local testing, will remove this route after the login page is complete

File: server/routes/index.ts (hunk 10 above)
Comment (zengyan-amazon) [reply server/routes/index.ts]: corrected

File: server/routes/index.ts (hunk 11 above)
Comment (zengyan-amazon) [reply server/routes/index.ts]: updated to use schema map
